proof/refine/RISCV64/EmptyFail.thy

```diff
@@ -12,38 +12,26 @@ begin
    Unless there is a good reason, they should all be [intro!, wp, simp] *)
 
 lemma empty_fail_projectKO [simp, intro!]:
-  "empty_fail (projectKO v)"
-  unfolding empty_fail_def projectKO_def
-  by (simp add: return_def fail_def split: option.splits)
+  "empty_fail (gets_the $ projectKO v)" by wpsimp
```

Comment:
This looks unlikely to apply with `simp`, because the `$` would usually already be gone when this rule is used. I think a general lemma `empty_fail (gets_the f)` (for any f) would be better here.

Reply:
Hm, but that would not actually solve a `empty_fail (gets_the $ _)` goal. So maybe two rules, the one I suggested above with `[simp, intro!]`, and `empty_fail (gets_the $ _)` with just `[intro!]`.

Reply:
We should do this properly: if `$` does occur in terms with `empty_fail`, we should have a rule for `$`, then we never need any combination rules. So we should have `empty_fail (f x) ==> empty_fail (f $ x)` as `[intro!]`, and then no other rules with `$` in that position, i.e. only `empty_fail (gets_the f)` as `[simp, intro!]`.

Reply:
We do have `gets_the_empty_fail` which is generated by a `crunch` in `EmptyFail_AI`. It is `[wp]` only. Nothing else in that file seems to have other attributes. Should I remove `empty_fail_projectKO`, and then just add the `empty_fail (f x) ==> empty_fail (f $ x)` rule as `[intro!]`, and that should do?

Reply:
Yes, that sounds good. I can't guarantee that we haven't missed anything, but I think it should work.